src/gf.c

```diff
@@ -937,8 +936,10 @@ DLLEXPORT jl_function_t *jl_instantiate_staged(jl_methlist_t *m, jl_tupletype_t
         }
         ex = oldast;
     }
-    func = (jl_function_t*)jl_toplevel_eval_in(m->func->linfo->module, (jl_value_t*)ex);
-    func->linfo->name = m->func->linfo->name;
+    jl_lambda_info_t *li = (jl_lambda_info_t*)jl_expand_in(m->func->linfo->module, (jl_value_t*)ex);
+    assert(jl_is_lambda_info(li));
+    li->name = m->func->linfo->name;
+    func = jl_new_closure(NULL, (jl_value_t*)jl_emptysvec, li);
```

Comment:
@JeffBezanson i think this is what you were suggesting for replacing the `eval`? what does the second argument to `jl_new_closure` mean in this case (i copied this from interpreter.c)?